core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -201,112 +201,121 @@ private void cleanExpiredFiles(List<Snapshot> snapshots, Set<Long> validIds, Set
+                throw new RuntimeIOException(e,
+                    "Failed to close manifest list: %s", snapshot.manifestListLocation());
+              }
+            });
 
     // find manifests to clean up that were only referenced by snapshots that have expired
     Set<String> manifestListsToDelete = Sets.newHashSet();
     Set<String> manifestsToDelete = Sets.newHashSet();
     Set<ManifestFile> manifestsToRevert = Sets.newHashSet();
-    for (Snapshot snapshot : base.snapshots()) {
-      long snapshotId = snapshot.snapshotId();
-      if (!validIds.contains(snapshotId)) {
-        // determine whether the changes in this snapshot are in the current table state
-        if (pickedAncestorSnapshotIds.contains(snapshotId)) {
-          // this snapshot was cherry-picked into the current table state, so skip cleaning it up. its changes will
-          // expire when the picked snapshot expires.
-          // A -- C -- D (source=B)
-          //  `- B <-- this commit
-          continue;
-        }
-
-        long sourceSnapshotId = PropertyUtil.propertyAsLong(
-            snapshot.summary(), SnapshotSummary.SOURCE_SNAPSHOT_ID_PROP, -1);
-        if (ancestorIds.contains(sourceSnapshotId)) {
-          // this commit was cherry-picked from a commit that is in the current table state. do not clean up its
-          // changes because it would revert data file additions that are in the current table.
-          // A -- B -- C
-          //  `- D (source=B) <-- this commit
-          continue;
-        }
-
-        if (pickedAncestorSnapshotIds.contains(sourceSnapshotId)) {
-          // this commit was cherry-picked from a commit that is in the current table state. do not clean up its
-          // changes because it would revert data file additions that are in the current table.
-          // A -- C -- E (source=B)
-          //  `- B `- D (source=B) <-- this commit
-          continue;
-        }
-
-        // find any manifests that are no longer needed
-        try (CloseableIterable<ManifestFile> manifests = readManifestFiles(snapshot)) {
-          for (ManifestFile manifest : manifests) {
-            if (!validManifests.contains(manifest.path())) {
-              manifestsToDelete.add(manifest.path());
-
-              boolean isFromAncestor = ancestorIds.contains(manifest.snapshotId());
-              boolean isFromExpiringSnapshot = expiredIds.contains(manifest.snapshotId());
-
-              if (isFromAncestor && manifest.hasDeletedFiles()) {
-                // Only delete data files that were deleted in by an expired snapshot if that
-                // snapshot is an ancestor of the current table state. Otherwise, a snapshot that
-                // deleted files and was rolled back will delete files that could be in the current
-                // table state.
-                manifestsToScan.add(manifest.copy());
+    Tasks.foreach(base.snapshots()).noRetry().suppressFailureWhenFinished()
```

Comment:
Same here. Wouldn't it be better to retry?

Reply:
Done.